include/RAJA/policy/cuda/reduce.hpp

```diff
@@ -1040,30 +865,76 @@ struct ReduceAtomic_Data {
   //  free device pointers
   bool teardownForDevice()
   {
-    bool act = own_device_ptr;
+    bool act = owns_device_pointer;
     if (act) {
-      device_mempool_type::getInstance().free(device);
+      data_mempool_type::getInstance().free(device);
       device = nullptr;
-      device_zeroed_mempool_type::getInstance().free(device_count);
+      count_mempool_type::getInstance().free(device_count);
       device_count = nullptr;
-      own_device_ptr = false;
+      owns_device_pointer = false;
     }
     return act;
   }
 };
 
+
 //! Cuda Reduction entity -- generalize on reduction, and type
-template <typename Combiner, typename T, bool maybe_atomic>
+template <typename Combiner, typename T, typename tuning>
 class Reduce
 {
+  static constexpr size_t replication = (tuning::replication > 0)
+      ? tuning::replication
+      : 1;
+  static constexpr size_t atomic_stride = (tuning::atomic_stride > 0)
+      ? tuning::atomic_stride
+      : ((policy::cuda::ATOMIC_DESTRUCTIVE_INTERFERENCE_SIZE > sizeof(T))
+        ? RAJA_DIVIDE_CEILING_INT(policy::cuda::ATOMIC_DESTRUCTIVE_INTERFERENCE_SIZE, sizeof(T))
+        : 1);
+
+  using Accessor = std::conditional_t<(tuning::comm_mode == block_communication_mode::block_fence),
+      impl::AccessorDeviceScopeUseBlockFence,
+      std::conditional_t<(tuning::comm_mode == block_communication_mode::device_fence),
+        impl::AccessorDeviceScopeUseDeviceFence,
+        void>>;
+
+  static constexpr bool atomic_policy =
+      (tuning::algorithm == reduce_algorithm::init_device_combine_atomic_block) ||
+      (tuning::algorithm == reduce_algorithm::init_host_combine_atomic_block);
+  static constexpr bool atomic_available = RAJA::reduce::cuda::cuda_atomic_available<T>::value;
+
+  //! cuda reduction data storage class and folding algorithm
+  using reduce_data_type = std::conditional_t<(tuning::algorithm == reduce_algorithm::combine_last_block) ||
+                                              (atomic_policy && !atomic_available),
+      cuda::ReduceLastBlock_Data<Combiner, Accessor, T, replication, atomic_stride>,
+      std::conditional_t<atomic_available,
+        std::conditional_t<(tuning::algorithm == reduce_algorithm::init_device_combine_atomic_block),
+          cuda::ReduceAtomicDeviceInit_Data<Combiner, Accessor, T, replication, atomic_stride>,
+          std::conditional_t<(tuning::algorithm == reduce_algorithm::init_host_combine_atomic_block),
+            cuda::ReduceAtomicHostInit_Data<Combiner, T, replication, atomic_stride>,
+            void>>,
+        void>>;
+
+  static constexpr size_t tally_slots = reduce_data_type::tally_slots;
+
+  using TallyType = PinnedTally<T, tally_slots, typename reduce_data_type::tally_mempool_type>;
```

Comment:
Here is the logic used by the reducer to decide which implementation to use based on the information in the reducer tuning object.

Reply:
Add that comment to the code documentation?